core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -545,9 +551,20 @@ class LogManager(logDirs: Array[File],
         if (!isNew && offlineLogDirs.nonEmpty)
           throw new KafkaStorageException(s"Can not create log for $topicPartition because log directories ${offlineLogDirs.mkString(",")} are offline")
 
-        val dataDir = nextLogDir()
+        val logDir = {
+          val preferredLogDir = preferredLogDirs.get(topicPartition)
+          if (preferredLogDir != null)
+            preferredLogDir
+          else
+            nextLogDir().getAbsolutePath
+        }
+        if (!isLogDirOnline(logDir))
+          throw new KafkaStorageException(s"Can not create log for $topicPartition because log directory $logDir is offline")
+        // Do not re-use the user-specified log directory if the replica is deleted and re-created in the future.
```

Comment:
Good point. I have updated the comment as suggested. And it is only removed after the log has been successfully created.